clang/lib/AST/ASTContext.cpp

```diff
@@ -12037,8 +12037,12 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
     if (LangOpts.SYCLIsDevice && isa<CXXMethodDecl>(D)) {
       if (auto *A = D->getAttr<SYCLDeviceIndirectlyCallableAttr>())
         return !A->isImplicit();
+      // Implicit 'sycl_device' attribute is treated as explicit one if method
+      // is also annotated with 'add_ir_attributes_function' attribute, because
+      // the latter can work as an alias for SYCL_EXTERNAL.
       if (auto *A = D->getAttr<SYCLDeviceAttr>())
```

Comment:
Do you know why we have the Implicit check? Why don't we emit functions irrespective of whether the attribute is implicit or explicit?

Reply:
> Do you know why we have the Implicit check? Why don't we emit functions irrespective of whether the attribute is implicit or explicit?

No, I don't. `git blame` just goes years back to some refactoring commit and there is not much info about why we specifically look for an explicit attribute (3baec18733227).

Looking at the `clang` codebase, there are only 3 places where we create implicit `sycl_device` attribute:
- `SemaLookup.cpp` when we generate declarations for SPIR-V friendly IR built-ins
- `SemaDeclAttr.cpp` when we handle legacy indirectly callable attribute we have
- `SemaDeclAttr.cpp` when we handle `add_ir_function_attributes` attribute (my recent patch)

Let me try and just drop requirement for the attribute to be explicit and see if anything breaks.

Reply:
Both `check-clang` and `check-sycl` pass if I remove `!A->isImplicit()` check for `sycl_device` attribute. I still have no idea why it was introduced in the first place. Perhaps it is some kind of  legacy? Before 3baec18 we used to add the attribute everywhere alongside call graph (implicitly) to indicate which functions should be emitted and which are not.

Considering that now there is only a few places where we implicitly add the attribute and only a couple of them is relevant to `CXXMethodDecl`s, what do you think about just dropping the requirement for the attribute to be explicit?

Reply:
I think it makes sense dropping the `isImplicit `check. In the cases mentioned above it feels like we actually want the functions/builtins to be emitted? I don't know what the purpose of adding this attribute is, if we don't want to emit the function.